File: 2021/day02.cpp
// https://adventofcode.com/2021/day/2

#include "../common/common.h"

#include <algorithm>
#include <array>
#include <iterator>
#include <numeric>
#include <print>
#include <ranges>
#include <string>
#include <string_view>
#include <vector>

struct command_t {
  std::string command;
  int number;
};

using commands_t = std::vector<command_t>;

commands_t parse(const std::string& filename) {
  commands_t commands;
  for (std::string_view line : aoc::views::read_lines(filename)) {
    auto values = aoc::split_once(line, ' ');
    commands.emplace_back(std::string{values[0]},
                          aoc::to_number<int>(values[1]));
  }
  return commands;
}

int solve_part1(const commands_t& commands) {
  int horizontal = 0;
  int depth = 0;

  for (const auto& [command, number] : commands) {
    if (command == "forward") {
      horizontal += number;
    } else if (command == "down") {
      depth += number;
    } else if (command == "up") {
      depth -= number;
    }
  }

  return (horizontal * depth);
}

int solve_part2(const commands_t& commands) {
  int horizontal = 0;
  int depth = 0;
  int aim = 0;

  for (const auto& [command, number] : commands) {
    if (command == "forward") {
      horizontal += number;
      depth += (aim * number);
    } else if (command == "down") {
      aim += number;
    } else if (command == "up") {
      aim -= number;
    }
  }

  return (horizontal * depth);
}

int main() {
  std::println("Part 1");
  const auto example = parse("day02.example");
  AOC_EXPECT_RESULT(150, solve_part1(example));
  const auto input = parse("day02.input");
  AOC_EXPECT_RESULT(2120749, solve_part1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(900, solve_part2(example));
  AOC_EXPECT_RESULT(2138382217, solve_part2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2021/src/day02.rs
type Direction = (i32, i32);

fn parse(filename: &str) -> Vec<Direction> {
    aoc::file::read_lines(filename)
        .iter()
        .map(|line| {
            let (command, value_str) = line.split_once(' ').unwrap();
            let value = value_str.parse::<i32>().unwrap();
            match command {
                "forward" => (value, 0),
                "down" => (0, value),
                "up" => (0, -value),
                _ => unreachable!("Invalid command"),
            }
        })
        .collect::<Vec<_>>()
}

fn solve_case1(directions: &[Direction]) -> u32 {
    let mut position = (0, 0);
    for direction in directions {
        position.0 += direction.0;
        position.1 += direction.1;
    }
    (position.0 as u32) * (position.1 as u32)
}

fn solve_case2(directions: &[Direction]) -> u32 {
    let mut position = (0, 0);
    let mut aim = 0;
    for (forward, aim_dir) in directions {
        if (*forward > 0) {
            position.0 += *forward;
            position.1 += aim * *forward;
        } else {
            aim += *aim_dir;
        }
    }
    (position.0 as u32) * (position.1 as u32)
}

fn main() {
    println!("Part 1");
    let example = parse("day02.example");
    assert_eq!(150, solve_case1(&example));
    let input = parse("day02.input");
    assert_eq!(2120749, solve_case1(&input));

    println!("Part 2");
    assert_eq!(900, solve_case2(&example));
    assert_eq!(2138382217, solve_case2(&input));
}
